File: src/main/scala/com/nekokittygames/Thaumic/Tinkerer/common/items/ItemMobAspect.scala
package com.nekokittygames.Thaumic.Tinkerer.common.items

import java.util

import com.nekokittygames.Thaumic.Tinkerer.common.core.misc.ItemNBT
import com.nekokittygames.Thaumic.Tinkerer.common.libs.{LibMisc, LibNames, LibItemNames}
import net.minecraft.client.resources.model.ModelResourceLocation
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.{Item, ItemStack}
import net.minecraftforge.fml.relauncher.{Side, SideOnly}
import net.minecraftforge.oredict.OreDictionary
import thaumcraft.api.ThaumcraftApiHelper
import thaumcraft.api.aspects.Aspect

/**
 * Created by katsw on 09/10/2015.
 */
object ItemMobAspect extends ModItem{

  final var ASPECT_TAG="aspect"
  final var INFUSED_TAG="infused"
  final var CONDENSED_TAG="condensed"

  setUnlocalizedName(LibItemNames.MOB_ASPECT)
  setHasSubtypes(true)


  override def addInformation(stack: ItemStack, playerIn: EntityPlayer, tooltip: util.List[String], advanced: Boolean): Unit =
    {
      super.addInformation(stack, playerIn, tooltip.asInstanceOf[java.util.List[String]], advanced)
      val aspect:Aspect=getAspect(stack)
      if(aspect!=null)
        tooltip.asInstanceOf[java.util.List[String]].add(lang.format("mobAspects.aspect.info",aspect.getName))
      if(isCondensed(stack))
        tooltip.asInstanceOf[java.util.List[String]].add(lang.translate("mobaspects.condensed"))
      if(isInfused(stack))
        tooltip.asInstanceOf[java.util.List[String]].add(lang.translate("mobaspects.infused"))
    }


  def getAspect(itemStack: ItemStack): Aspect =
  {
    Aspect.getAspect(ItemNBT.getItemStackTag(itemStack).getString(ASPECT_TAG))
  }
  def setAspect(itemStack: ItemStack,aspect:Aspect) =
  {
      ItemNBT.getItemStackTag(itemStack).setString(ASPECT_TAG,aspect.getTag)
  }

  def isInfused(itemStack: ItemStack):Boolean=
  {
      ItemNBT.getItemStackTag(itemStack).getBoolean(INFUSED_TAG)
  }

  def setInfused(itemStack: ItemStack,infused: Boolean): Unit =
  {
    ItemNBT.getItemStackTag(itemStack).setBoolean(INFUSED_TAG,infused)
  }

  def isCondensed(itemStack: ItemStack):Boolean=
  {
    ItemNBT.getItemStackTag(itemStack).getBoolean(CONDENSED_TAG)
  }

  def setCondensed(itemStack: ItemStack,condensed: Boolean): Unit =
  {
    ItemNBT.getItemStackTag(itemStack).setBoolean(CONDENSED_TAG,condensed)
  }

  override def getSubItems(itemIn: Item, tab: CreativeTabs, subItems: util.List[ItemStack]): Unit =
    {
      super.getSubItems(itemIn, tab, subItems)
      Aspect.getPrimalAspects.toArray.foreach
      {
        aspect=>
          val stack:ItemStack=new ItemStack(ItemMobAspect)
          setAspect(stack,aspect.asInstanceOf[Aspect])
          subItems.asInstanceOf[java.util.List[ItemStack]].add(stack)
      }
      Aspect.getCompoundAspects.toArray.foreach
      {
        aspect=>
          val stack:ItemStack=new ItemStack(ItemMobAspect)
          setAspect(stack,aspect.asInstanceOf[Aspect])
          subItems.asInstanceOf[java.util.List[ItemStack]].add(stack)
      }
    }

  override def getColorFromItemStack(stack: ItemStack, renderPass: Int): Int =
    {
      val aspect:Aspect=getAspect(stack)
      if(aspect!=null)
        aspect.getColor
      else
        0
    }

  override def getUnlocalizedName(stack: ItemStack): String =
    {
      if(isInfused(stack))
        return super.getUnlocalizedName(stack)+".infused"
      if(isCondensed(stack))
        return super.getUnlocalizedName(stack)+".condensed"
      super.getUnlocalizedName(stack)
    }

  override def hasEffect(stack: ItemStack): Boolean =
    {
      if(isInfused(stack))
        true
      else
        super.hasEffect(stack)

    }


  def doesMobAspectMatchForInfusion(stack0:ItemStack,aspect:Aspect):Boolean= {
    if (stack0.isInstanceOf[ItemStack]) {
      if(stack0.getItem!=ItemMobAspect)
        false
      else
        {
          if(ItemMobAspect.getAspect(stack0).getTag.equalsIgnoreCase(aspect.getTag) && ItemMobAspect.isCondensed(stack0))
            true
          else
            false
        }
    }
    else
      false

  }

  @SideOnly(Side.CLIENT)
  override def getModel(stack: ItemStack, player: EntityPlayer, useRemaining: Int): ModelResourceLocation =
    {
      if(isCondensed(stack) || isInfused(stack))
        {
         new ModelResourceLocation(LibMisc.MOD_ID+":"+LibItemNames.MOB_ASPECT+"Condensed","inventory")
        }
      else
        new ModelResourceLocation(LibMisc.MOD_ID+":"+LibItemNames.MOB_ASPECT,"inventory")
    }
}

File: src/main/java/com/nekokittygames/thaumictinkerer/common/items/ItemMobAspect.java
/*
 * Copyright (c) 2020. Katrina Knight
 */

package com.nekokittygames.thaumictinkerer.common.items;

import com.nekokittygames.thaumictinkerer.ThaumicTinkerer;
import com.nekokittygames.thaumictinkerer.common.helper.IItemVariants;
import com.nekokittygames.thaumictinkerer.common.libs.LibItemNames;
import net.minecraft.client.resources.I18n;
import net.minecraft.client.util.ITooltipFlag;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.NonNullList;
import net.minecraft.world.World;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;
import thaumcraft.api.aspects.Aspect;

import javax.annotation.Nullable;
import java.util.List;
import java.util.Set;
import java.util.SortedSet;
import java.util.TreeSet;

public class ItemMobAspect extends TTItem implements IItemVariants {
    public static String ASPECT_NAME="aspectName";
    public ItemMobAspect() {
        super(LibItemNames.MOB_ASPECT);
    }

    @Override
    protected CreativeTabs getModCreativeTab() {
        return ThaumicTinkerer.getTabAspects();
    }

    @Override
    public String GetVariant(ItemStack stack) {
        return stack.getTagCompound() != null ? stack.getTagCompound().getString(ASPECT_NAME) : "aer";
    }

    @Override
    public Set<String> GetVariants() {
        return Aspect.aspects.keySet();
    }

    @Override
    @SideOnly(Side.CLIENT)
    public void addInformation(ItemStack stack, @Nullable World worldIn, List<String> tooltip, ITooltipFlag flagIn) {
        super.addInformation(stack, worldIn, tooltip, flagIn);
        if(stack.hasTagCompound() && stack.getTagCompound().hasKey(ASPECT_NAME)) {
            Aspect aspect=Aspect.getAspect(stack.getTagCompound().getString(ASPECT_NAME));
            if(aspect!=null) {
                tooltip.add(I18n.format("thaumictinkerer.mobaspect.type", aspect.getName()));
            }
            else
            {
                tooltip.add(I18n.format("thaumictinkerer.mobaspect.invalid"));
            }
        }
        else {
            tooltip.add(I18n.format("thaumictinkerer.mobaspect.invalid"));
        }
    }

    @Override
    public void getSubItems(CreativeTabs tab, NonNullList<ItemStack> items) {
        if (this.isInCreativeTab(tab))
        {
            SortedSet<String> sortedAspects = new TreeSet<>(GetVariants());
            for(String aspect:sortedAspects) {
                ItemStack itemStack=new ItemStack(this);
                NBTTagCompound cmp=itemStack.getTagCompound();
                if(cmp==null)
                    cmp=new NBTTagCompound();
                cmp.setString(ASPECT_NAME,aspect);
                itemStack.setTagCompound(cmp);
                items.add(itemStack);
            }
        }
    }
}
